service/history/workflow/task_refresher.go

```diff
@@ -140,33 +173,44 @@ func (r *TaskRefresherImpl) RefreshTasks(
 		ctx,
 		mutableState,
 		taskGenerator,
+		minVersionedTransition,
 	); err != nil {
 		return err
 	}
 
 	if err := r.refreshTasksForWorkflowSearchAttr(
 		mutableState,
 		taskGenerator,
+		minVersionedTransition,
 	); err != nil {
 		return err
 	}
 
 	return r.refreshTasksForSubStateMachines(
 		mutableState,
+		minVersionedTransition,
 	)
 }
 
 func RefreshTasksForWorkflowStart(
 	ctx context.Context,
 	mutableState MutableState,
 	taskGenerator TaskGenerator,
+	minVersionedTransition *persistencespb.VersionedTransition,
 ) error {
 
 	executionState := mutableState.GetExecutionState()
 	if executionState.Status != enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING {
 		return nil
 	}
 
+	if CompareVersionedTransition(
```

Comment:
This is more intuitive to understand why it's needed here but I'd still explain the purpose of this check.